### DISCUSSION TITLE
Control of the preconditioning programs

### DISCUSSION THREAD

**Discussion by 111Sebu111** [flobz/psa_car_controller#693] [Ideas]
In the MyPeugeot app you are able to create a preconditioning program with a certain schedule. It would be great if the preconditioning program could be also maintained/controlled through this. I think I remember seeing that this should be possible with the API. 

In winter and with the car plugged in, it seems the preconditioning program runs preconditioning differently that just simple manual activation. You are able to preheat the car better.

**Comment by sod75** [comment]
Not sure it's useful to embed a scheduler in this project. You can easily schedule the API call elsewhere. Most have it integrated in Homeassistant, schedule it there. or if you're using it standalone use your OS to schedule it ?

**Reply by 111Sebu111** [reply]
I use Home Assistant as well. So I'm more looking on managing the preconditioning program schedules rather than preconditioning on/off (different API I believe). I might be wrong here, but this is how I think it works. So if I create a preconditioning program schedule with MyPeugeot app, it will upload that to the car. Then the car will know by the outside temperature, when it needs to be started. For example when there -10 degrees outside it will start around 40 minutes before the departure. If I run it manually, it will only run for 20 minutes or so. This more a need for the winter times and to be able to properly melt the snow and ice from windows.

**Reply by sod75** [reply]
Ah, I see. Didn't even know this existed.
Not sure how dynamic this is, but it should indeed be possible in the API to push a certain pre-programmed preconditioning

![image](https://github.com/flobz/psa_car_controller/assets/4673458/ab3770be-ef13-4844-9b36-4419c5305e4c)

had a quick look in the code and I do see references to this functionality
`
    self.precond_programs = {}`

```
                programs = data["precond_state"].get("programs", None)
                if programs:
                    self.precond_programs[data["vin"]] = data["precond_state"]["programs"]
```

```
    def preconditioning(self, vin, activate: bool):
        if activate:
            value = "activate"
        else:
            value = "deactivate"
        if vin in self.precond_programs:
            programs = self.precond_programs[vin]
        else:
            programs = DEFAULT_PRECONDITIONING_PROGRAM
        msg = self.mqtt_request(vin, {"asap": value, "programs": programs}, "/ThermalPrecond")
        logger.info("Preconditioning: %s", msg)
        self.publish(msg)
        return True
```

so maybe it's just a question of calling it with the right parameters...

**Reply by 111Sebu111** [reply]
Awesome. Thanks for having looking into it.

**Reply by sod75** [reply]
Some more info, 

From the Connected services FAQ

```
• Remote heating and air conditioning: Prepare your vehicle in advance by heating or cooling the
passenger compartment.
•
Heating and cooling uses a lot of energy. If you use this function while charging, it will not
affect the range of the vehicle and the vehicle will be ready for use at a comfortable
temperature.
•
You do not have to enter a temperature yourself, as the e-Remote Controls will always aim for
a comfortable temperature (21°C), heating when it is cold outside and using the air
conditioning when it is hot. When you program a preconditioning, the system starts between
30 and 45 minutes before the chosen time to make sure that your vehicle reaches a
comfortable temperature and remains active 10 minutes after. It is therefore the time at which
you will use your vehicle that must be scheduled.
```

So one should send the call with just a time similar to how we set the Charge time I suppose.

**Reply by 111Sebu111** [reply]
I think it is slightly more complicated than that.

I'm guessing there should be an API that create/maintain the preconditioning program. Something around that you define the time and days of occurrence (e.g. run on Mondays and Tuesdays). Reason why I'm thinking this, is that in the PeugeotApp you can add multiple preconditioning schedule.

Then a separate API to enable/disable the particular preconditioning program by referring to program 'slot'.

![image](https://github.com/flobz/psa_car_controller/assets/88944907/9375e20c-a2cf-4c3c-a694-b0351460f9fe)

**Comment by InJotec** [comment]
Hi, I'm using Home Assistant integration, everything works fine, but I miss this functionality of being able to modify, create, delete, activate or deactivate the conditioning presets. 
I miss it a lot because the official car app does not work in any way this section, I always get server error, the rest of the official app if it works, horribly slow, but not to modify the schedules. Unfortunately my car does not allow to edit these presets from the car itself, only from the app.

Is there any plan to add this functionality?
